wolfhsm/wh_transport_mem.h

```diff
@@ -126,6 +127,7 @@ typedef struct {
     int initialized;
     uint16_t req_size;
     uint16_t resp_size;
+    uint16_t* cancel_seq;
```

Comment:
The current cancel callback scheme is transport agnostic, `cancel_seq` is an out-of-band signal for shared memory by virtue of being a separate piece of memory. Implementations that don't use shared memory should set up their own out of band method for passing the current canceled sequence, note that the server calls the callback function directly expecting it's uint16_t pointer to be set but knows nothing with regard to how the implementation achieves this. Same for the client, `wh_CommClient_Cancel` simply calls the Cancel cb with the current sequence but doesn't care how the implementation achieves it. In other words, transport_mem is using shared mem to send the seq but the callers don't see or care how the transport layer sends the signal and other implementations may do it however is necessary.

Reply:
And I'll add that in the mem transport I simply set cancel_seq to 0 when it is not set to represent no signal but other implementations need to check their own transport and if there's no signal set 0 for the caller manually, might have been part of the confusion.

Reply:
> The current cancel callback scheme is transport agnostic, cancel_seq is an out-of-band signal for shared memory by virtue of being a separate piece of memory

when I say transport agnostic I mean not requiring modifying the transport layer at all (here you added a bunch of stuff to transport mem, for example).

> Implementations that don't use shared memory should set up their own out of band method for passing the current canceled sequence, note that the server calls the callback function directly expecting it's uint16_t pointer to be set but knows nothing with regard to how the implementation achieves this. Same for the client, wh_CommClient_Cancel simply calls the Cancel cb with the current sequence but doesn't care how the implementation achieves it.

I think the problem is that the state variable _does_ need to be shared between both sides as it is currently written, and that is what I am saying we can't do, since there is no built in way to enforce coherency on that piece of memory (which we would need to do for any region of memory outside the one sequential uncached shared buffer that transport mem uses). 

I'm suggesting that instead of making this a transport specific thing, we make it a port-specific thing entirely done through callbacks. That way you don't need to change anything about the TransportXXX at all. Instead, the client should just invoke a cancel callback registered through the config structure when a "cancel" client function is invoked. And there would be no shared state required. Then on the server side, the cancellation state can be stored locally in the server context, and set out-of-band (e.g. in a server ISR) through a function like `wh_Server_CancelRequest(whServer* server, ...)`. Then any cancellable operation like CMAC can poll the server cancellation state in a similar manner to how it is done now, but through the server context instead of the transport layer. Pretty much entirely mimicking the connection callback functionality. 

tl;dr: I think we should make the cancellation port-specific, not transport-specific,  with no shared state between server and client. This will dramatically simplify things and eliminate any concerns around shared memory access / caching/etc.